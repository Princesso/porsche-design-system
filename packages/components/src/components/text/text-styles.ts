import type { JssStyle } from 'jss';
import type { BreakpointCustomizable, TextAlign, TextColor, TextSize, TextWeight, Theme } from '../../types';
import { buildResponsiveStyles, getCss, textMap } from '../../utils';
import { addImportantToEachRule } from '../../styles';
import { textSmallStyle } from '@porsche-design-system/utilities-v2';
import { getEllipsisJssStyle, getSlottedTypographyJssStyle } from '../../styles/typography-styles';
import { getThemedTextColor } from '../../styles/text-icon-styles';
import { getFontWeight } from '../../styles/font-weight-styles';

export const getComponentCss = (
  size: BreakpointCustomizable<TextSize>,
  weight: TextWeight,
  align: TextAlign,
  color: TextColor,
  ellipsis: boolean,
  theme: Theme
): string => {
  // function is local to reuse `weight` parameter
  // TODO: font shorthand isn't really the best choice but we don't have any better alternative atm
  const getSizeJssStyle = (textSize: TextSize): JssStyle => {
    const fontWeightValue = getFontWeight(weight);
    return textSize === 'inherit'
      ? {
          fontSize: textSize,
          fontWeight: fontWeightValue,
        }
      : { font: textMap[textSize].font.replace('400', fontWeightValue) };
  };

  return getCss({
    '@global': {
      ':host': {
        display: 'block',
      },
      '::slotted': {
        '&(p),&(address),&(blockquote),&(figcaption),&(cite),&(time),&(legend)': addImportantToEachRule(
          getSlottedTypographyJssStyle()
        ),
      },
    },
    root: {
      display: 'inherit',
      padding: 0,
      margin: 0,
      textAlign: align,
      ...textSmallStyle,
      color: getThemedTextColor(theme, color),
      listStyleType: 'none',
      whiteSpace: 'inherit',
      ...(ellipsis && getEllipsisJssStyle()),
      ...buildResponsiveStyles(size, getSizeJssStyle),
    },
  });
};
