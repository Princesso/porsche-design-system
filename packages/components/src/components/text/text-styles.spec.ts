import { getComponentCss } from './text-styles';

describe('getComponentCss()', () => {
  it.each<Parameters<typeof getComponentCss>>([
    ['small', 'regular', 'left', 'primary', false, 'light'],
    ['small', 'regular', 'left', 'default', false, 'light'],
    ['x-large', 'thin', 'right', 'brand', true, 'dark'],
    ['x-large', 'regular', 'center', 'brand', false, 'dark'],
    ['large', 'semi-bold', 'left', 'notification-info', true, 'light'],
    ['large', 'semibold', 'left', 'notification-error', true, 'light'],
    ['medium', 'bold', 'right', 'neutral-contrast-high', true, 'dark'],
    ['x-small', 'semibold', 'left', 'inherit', false, 'light'],
    ['inherit', 'bold', 'center', 'default', false, 'light'],
    [
      { base: 'small', xs: 'large', s: 'medium', m: 'inherit', l: 'x-small', xl: 'x-large' },
      'bold',
      'center',
      'notification-error',
      true,
      'dark',
    ],
  ])(
    'should return correct css for size: %j, weight: %s, align: %s, color: %s, ellipsis: %o and theme: %s',
    (...args) => {
      expect(getComponentCss(...args)).toMatchSnapshot();
    }
  );
});
