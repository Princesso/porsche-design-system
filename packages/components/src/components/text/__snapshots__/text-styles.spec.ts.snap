// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for size: "inherit", weight: bold, align: center, color: default, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: center;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #010205;
  list-style-type: none;
  white-space: inherit;
  font-size: inherit;
  font-weight: 700;
}"
`;

exports[`getComponentCss() should return correct css for size: "large", weight: semi-bold, align: left, color: notification-info, ellipsis: true and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: left;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #1E5BEB;
  list-style-type: none;
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: clamp(1.44rem, .3vw + 1.38rem, 1.71rem);
  font-weight: 600;
}"
`;

exports[`getComponentCss() should return correct css for size: "large", weight: semibold, align: left, color: notification-error, ellipsis: true and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: left;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #E7323B;
  list-style-type: none;
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: clamp(1.44rem, .3vw + 1.38rem, 1.71rem);
  font-weight: 600;
}"
`;

exports[`getComponentCss() should return correct css for size: "medium", weight: bold, align: right, color: neutral-contrast-high, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: right;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #AFB0B3;
  list-style-type: none;
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: clamp(1.2rem, .12vw + 1.18rem, 1.31rem);
  font-weight: 700;
}"
`;

exports[`getComponentCss() should return correct css for size: "small", weight: regular, align: left, color: default, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: left;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #010205;
  list-style-type: none;
  white-space: inherit;
  font-size: 1rem;
  font-weight: 400;
}"
`;

exports[`getComponentCss() should return correct css for size: "small", weight: regular, align: left, color: primary, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: left;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #010205;
  list-style-type: none;
  white-space: inherit;
  font-size: 1rem;
  font-weight: 400;
}"
`;

exports[`getComponentCss() should return correct css for size: "x-large", weight: regular, align: center, color: brand, ellipsis: false and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: center;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #FBFCFF;
  list-style-type: none;
  white-space: inherit;
  font-size: clamp(1.73rem, .57vw + 1.61rem, 2.24rem);
  font-weight: 400;
}"
`;

exports[`getComponentCss() should return correct css for size: "x-large", weight: thin, align: right, color: brand, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: right;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #FBFCFF;
  list-style-type: none;
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: clamp(1.73rem, .57vw + 1.61rem, 2.24rem);
  font-weight: 400;
}"
`;

exports[`getComponentCss() should return correct css for size: "x-small", weight: semibold, align: left, color: inherit, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: left;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: currentColor;
  list-style-type: none;
  white-space: inherit;
  font-size: .875rem;
  font-weight: 600;
}"
`;

exports[`getComponentCss() should return correct css for size: {"base":"small","xs":"large","s":"medium","m":"inherit","l":"x-small","xl":"x-large"}, weight: bold, align: center, color: notification-error, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none;
}
::slotted(p), ::slotted(address), ::slotted(blockquote), ::slotted(figcaption), ::slotted(cite), ::slotted(time), ::slotted(legend) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
  letter-spacing: inherit !important;
}
.root {
  display: inherit;
  margin: 0;
  padding: 0;
  text-align: center;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  letter-spacing: normal;
  color: #CB3333;
  list-style-type: none;
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 1rem;
  font-weight: 700;
}
@media(min-width:480px) {
  .root {
    font-size: clamp(1.44rem, .3vw + 1.38rem, 1.71rem);
    font-weight: 700;
  }
}

@media(min-width:760px) {
  .root {
    font-size: clamp(1.2rem, .12vw + 1.18rem, 1.31rem);
    font-weight: 700;
  }
}

@media(min-width:1000px) {
  .root {
    font-size: inherit;
    font-weight: 700;
  }
}

@media(min-width:1300px) {
  .root {
    font-size: .875rem;
    font-weight: 700;
  }
}

@media(min-width:1760px) {
  .root {
    font-size: clamp(1.73rem, .57vw + 1.61rem, 2.24rem);
    font-weight: 700;
  }
}
"
`;
